tripal/tripal
# Fix the "Remove" button on field widgets

# Bug Fix

### Closes #1920
### Depends on #1921 (Merged)

### Tripal Version: 4.x

## Description
When editing an entity, the "Remove" button will remove a linked entity (e.g. remove analysis from project) in the Drupal table, but does not delete it from the underlying chado linker table. If instead it is changed to -- Select -- then it is removed from both the Drupal and Chado tables. This is because the "Remove" button removes the item from the form state so then Chado storage doesn't know it existed and doesn't know to delete it in Chado.

To resolve this problem, I am storing a few values in the form state on the very first creation of the form,
and then in the `massageFormValues()` function, we can check for records that were there when the form was first built, but have been eliminated since. If that is the case, then put a reference in the `$values` array which will inform chado storage to delete the record in chado.

This problem will exist only for fields which support cardinality greater than one. Fields that need the fix:
 - [x] ChadoAnalysisWidgetDefault
 - [x] ChadoArrayDesignWidgetDefault
 - [x] ChadoAssayWidgetDefault
 - [x] ChadoBiomaterialWidgetDefault
 - [x] ChadoContactWidgetDefault
 - [x] ChadoFeatureMapWidgetDefault
 - [x] ChadoFeatureWidgetDefault
 - [x] ChadoOrganismWidgetDefault
 - [x] ChadoProjectWidgetDefault
 - [x] ChadoProtocolWidgetDefault
 - [x] ChadoPubWidgetDefault
 - [x] ChadoStockWidgetDefault
 - [x] ChadoStudyWidgetDefault

These will probably be different
 - [x] ChadoDbxrefWidgetDefault
 - [x] ChadoPropertySelectWidgetDefault
 - [x] ChadoPropertyStringWidgetDefault
 - [x] ChadoPropertyWidgetDefault
 - [x] ChadoSynonymWidgetDefault

Changes not needed because cardinality is always 1
 - ChadoAdditionalTypeWidgetDefault 


## Testing?
~~(We need PR #1921 merged into this branch to test)~~ merged
1. Build a docker on the 4.x branch (to test update hook)
2. Import Genetic and Genomic type collections and run the job `drush trp-run-jobs --username=drupaladmin --root=/var/www/drupal/web`
3. switch to this branch `git checkout tv4g1-issue1920-fix-remove-button`
4. Run update hook `drush updatedb`
```
>  [notice] Update started: tripal_chado_update_10406
>  [notice] Update completed: tripal_chado_update_10406
>  [notice] Message: Updating properties for Drupal table "tripal_entity__qtl_synonym", field  
> "qtl_synonym"
> 
>  [notice] Message: Updating properties for Drupal table  
> "tripal_entity__sequence_variant_synonym", field "sequence_variant_synonym"
> 
>  [notice] Message: Updating properties for Drupal table "tripal_entity__genetic_marker_synonym",  
> field "genetic_marker_synonym"
> 
>  [notice] Message: Updating properties for Drupal table  
> "tripal_entity__phenotypic_marker_synonym", field "phenotypic_marker_synonym"
> 
>  [notice] Message: Updating properties for Drupal table "tripal_entity__gene_synonym", field  
> "gene_synonym"
> 
>  [notice] Message: Updating properties for Drupal table "tripal_entity__mrna_synonym", field  
> "mrna_synonym"
> 
>  [notice] Message: Updating properties for Drupal table "tripal_entity__dna_library_synonym",  
> field "dna_library_synonym"
> 
>  [notice] Message: Updated 7 properties
> 
 [success] Finished performing updates.
```
5. Create three contacts
6. Create an organism
7. Create a "Genetic Marker", and a) link all of the contacts, b) add some "Database Reference Annotations", c) add some synonyms, and save.
![1958_1](https://github.com/user-attachments/assets/59d94c64-68f4-4269-bdac-b4c5f94c9c22)

8. Edit the Genetic Marker you created. Click the "Remove" button for one of the contacts, select "-- Select --" for another, then save.
![1958_2](https://github.com/user-attachments/assets/75784590-1725-4a56-b763-1b2236cf7540)

9. Contacts deleted using either of these methods should no longer appear on the entity.
![1958_3](https://github.com/user-attachments/assets/d8a0861b-cfad-4956-baca-ae05c0fffb50)

10. The two contacts should also be removed from the `project_contact` table and only one should remain - This is what was **not** happening before.
```
sitedb=> SELECT * FROM chado.feature_contact;
 feature_contact_id | feature_id | contact_id 
--------------------+------------+------------
                  7 |          1 |          4
(1 row)
```
11. Edit again and do the same for synonyms, click "Remove" on one, empty the synonym box for another and save. Check the `feature_synonym` table.
12. And likewise for the "Database Reference Annotations". Check the `feature_dbxref` table.
13. To test properties, import a gff3 file, for example [issue1824.gff3.txt](https://github.com/tripal/tripal/files/15109905/issue1824.gff3.txt)
14. Add new fields to "Gene"
15. publish Gene
16. Test removing some properties, you can examine with
`select * from featureprop FP left join cvterm T on FP.type_id=T.cvterm_id where FP.feature_id = (select feature_id from feature where name='PAU8');`
